docs/rst/knora-api-server/design-documentation/triplestore-updates.rst

```diff
@@ -202,35 +216,62 @@ a time can update a given resource. Before each update, the application
 constraint.
 
-Second, on GraphDB, the triplestore's consistency checker can be turned on to
-ensure that each update transaction respects the consistency constraints, as
-described in the section `Consistency checks`_ of the GraphDB documentation.
-This makes it possible to catch consistency constraint violations caused by
-bugs in Knora, and it also checks data that is uploaded directly into the
-triplestore without going through the Knora API. However, this feature is not
-yet enabled, because of problems described in `issue 33`_.
+However, the SPARQL update itself is our only chance to do pre-update checks
+in the same transaction that will perform the update. The design of the
+`SPARQL 1.1 Update`_ standard makes it possible to ensure that if certain
+conditions are not met, the update will not be performed. In our SPARQL update
+code, each update contains a ``WHERE`` clause, possibly a ``DELETE`` clause,
+and an ``INSERT`` clause. The ``WHERE`` clause is executed first. It performs
+consistency checks and provides values for variables that are used in the
+``DELETE`` and/or ``INSERT`` clauses. In our updates, if the expectations of
+the ``WHERE`` clause are not met (e.g. because the data to be updated does not
+exist), the ``WHERE`` clause should return no results; as a result, the update
+will not be performed.
+
+Regardless of whether the update succeeds or not, it returns nothing. So the
+only way to find out whether it was successful is to do a ``SELECT``
+afterwards. Moreover, if the update failed, there is no straightforward way to
+find out why. This is one reason why Knora does pre-update checks by means of
+separate ``SELECT`` queries, *before* performing the update. This makes it
+possible to return specific error messages to the user to indicate why an
+update cannot be performed.
+
+Moreover, while some checks are easy to do in a SPARQL update, others are
+difficult, impractical, or impossible. Easy checks include checking whether a
+resource or value exists or is deleted, and checking that the
+``knora-base:objectClassConstraint`` of a predicate matches the ``rdf:type`` of
+its intended object. Cardinality checks are not very difficult, but they perform
+poorly on Jena. Knora does not do permission checks in SPARQL, because its
+permission-checking algorithm is too complex to be implemented in SPARQL. For
+this reason, Knora's check for duplicate values cannot be done in SPARQL
+update code, because it relies on permission checks.
+
+
+GraphDB's consistency checker can be turned on to ensure that each update
+transaction respects the consistency constraints, as described in the section
+`Consistency checks`_ of the GraphDB documentation. This makes it possible to
+catch consistency constraint violations caused by bugs in Knora, and it also
+checks data that is uploaded directly into the triplestore without going
+through the Knora API. However, this feature is not yet enabled, because of
+problems described in `issue 33`_.
```

Comment:
> However, this feature is not yet enabled, because of problems described in issue 33.

This should be probably updated, to reflect that CC is now turned on.